qutebrowser/completion/models/sortfilter.py

```diff
@@ -140,21 +133,33 @@ def filterAcceptsRow(self, row, parent):
         if parent == QModelIndex() or not self.pattern:
             return True
 
-        try:
-            return self.srcmodel.custom_filter(self.pattern, row, parent)
-        except NotImplementedError:
-            for col in self.srcmodel.columns_to_filter:
-                idx = self.srcmodel.index(row, col, parent)
-                if not idx.isValid():
-                    # No entries in parent model
-                    continue
-                data = self.srcmodel.data(idx)
-                if not data:
-                    continue
-                elif self.pattern_re.search(data):
-                    return True
+        # first check the cache
```

Comment:
First I was trying to speed up the new features (first two commits) and admittedly caching didn't help as much as https://github.com/The-Compiler/qutebrowser/pull/1654/commits/ce46adde19f0b1d9e961c42a0bd31e116b0060b5, but I think it's still useful for the empty-match case mentioned above. It's still not ideal that writing a long search term throttles the CPU to 100% because Qt still filters in the background, but that could be solved only by temporarily disabling the filter as soon as the pattern does not match. No idea if that would work though...

Note that even the algorithmic complexity has increased, since #1651 means that we have to match each term from the beginning of the string instead of matching from the previous term to the end. So until we make some benchmark, it is very likely that this PR is slower than the original. We could optimize the implementation, perhaps even below the original performance level, but disabling the features would in theory always improve performance regardless of the implementation.

As for the benchmarking itself, first we need to have something to benchmark. If we measure just the inner method (`filterAcceptsRow`), there will be considerable overhead due to thousands of calls. Could we somehow hack into Qt and get signals when the filtering starts/ends?